File: public/static/portfolio.js
/* 포트폴리오 설계 모드: 상품 구성 → 비교 시뮬레이션
   API: GET /api/product-types, POST /api/simulate-portfolio (docs/07 참고) */

let PRODUCT_TYPES = {};
let lastResult = null;
let chartNet = null;
let chartAssets = null;

function formatWon(value) {
  const n = Math.round(Number(value) || 0);
  const sign = n < 0 ? "-" : "";
  const abs = Math.abs(n);
  const eok = Math.floor(abs / 100_000_000);
  const man = Math.floor((abs % 100_000_000) / 10_000);
  const rest = abs % 10_000;
  let parts = [];
  if (eok) parts.push(eok.toLocaleString("ko-KR") + "억");
  if (man) parts.push(man.toLocaleString("ko-KR") + "만");
  if (!eok && !man) return sign + abs.toLocaleString("ko-KR") + " 원";
  if (rest) parts.push(rest.toLocaleString("ko-KR"));
  return sign + parts.join(" ") + " 원";
}

const num = (id) => Number(document.getElementById(id).value) || 0;

// ── 금액칸 한글 단위 표시 (id 기반 큰 입력칸) ──────────────
function attachMoneyDisplays() {
  document.querySelectorAll("input.money").forEach((input) => {
    if (input.dataset.bound) return;
    input.dataset.bound = "1";
    const display = document.createElement("p");
    display.className = "money-display";
    input.insertAdjacentElement("afterend", display);
    const update = () => {
      const v = Number(input.value);
      display.textContent = v > 0 ? "= " + formatWon(v) : "";
    };
    input.addEventListener("input", update);
    update();
  });
}

// ── 상품 보유 행 ──────────────────────────────────────────
function buildTypeOptions(select) {
  select.innerHTML = "";
  for (const [code, meta] of Object.entries(PRODUCT_TYPES)) {
    const opt = document.createElement("option");
    opt.value = code;
    opt.textContent = meta.label;
    select.appendChild(opt);
  }
}

function addHolding(holdingsEl, preset = {}) {
  const tpl = document.getElementById("holding-template");
  const row = tpl.content.firstElementChild.cloneNode(true);
  const typeSel = row.querySelector(".h-type");
  const labelInput = row.querySelector(".h-label");
  const amountInput = row.querySelector(".h-amount");
  const yieldInput = row.querySelector(".h-yield");
  const incomeSpan = row.querySelector(".h-income");
  const noteEl = document.createElement("p");
  noteEl.className = "h-note";

  buildTypeOptions(typeSel);
  if (preset.product_type) typeSel.value = preset.product_type;
  if (preset.label) labelInput.value = preset.label;
  if (preset.amount) amountInput.value = preset.amount;
  if (preset.annual_yield_pct != null) yieldInput.value = preset.annual_yield_pct;

  const refresh = () => {
    const income = Math.round((Number(amountInput.value) || 0) * (Number(yieldInput.value) || 0) / 100);
    incomeSpan.textContent = "연 " + formatWon(income);
    noteEl.textContent = (PRODUCT_TYPES[typeSel.value] || {}).note || "";
    updatePortfolioSubtotal(holdingsEl.closest(".portfolio-card"));
  };
  typeSel.addEventListener("change", refresh);
  amountInput.addEventListener("input", refresh);
  yieldInput.addEventListener("input", refresh);
  row.querySelector(".h-del").addEventListener("click", () => {
    const card = holdingsEl.closest(".portfolio-card");
    row.remove();
    noteEl.remove();
    updatePortfolioSubtotal(card);
  });

  holdingsEl.appendChild(row);
  holdingsEl.appendChild(noteEl);
  refresh();
}

function updatePortfolioSubtotal(card) {
  if (!card) return;
  let total = 0;
  card.querySelectorAll(".holding-row").forEach((r) => {
    total += Number(r.querySelector(".h-amount").value) || 0;
  });
  card.querySelector(".pf-subtotal").textContent = "평가액 합계: " + formatWon(total);
}

// ── 포트폴리오 카드 ───────────────────────────────────────
let portfolioCount = 0;

function addPortfolio(preset = {}) {
  if (document.querySelectorAll(".portfolio-card").length >= 3) return;
  portfolioCount++;
  const card = document.createElement("div");
  card.className = "portfolio-card";
  card.innerHTML = `
    <div class="pf-head">
      <input class="pf-name" type="text" value="${preset.label || "포트폴리오 " + portfolioCount}">
      <button class="pf-del" title="이 포트폴리오 삭제">✕ 삭제</button>
    </div>
    <div class="holdings"></div>
    <div class="pf-foot">
      <button class="add-holding add-btn">＋ 상품 추가</button>
      <span class="pf-subtotal"></span>
    </div>`;
  const holdingsEl = card.querySelector(".holdings");
  card.querySelector(".add-holding").addEventListener("click", () => addHolding(holdingsEl));
  card.querySelector(".pf-del").addEventListener("click", () => {
    card.remove();
    toggleAddPortfolioBtn();
  });

  document.getElementById("portfolios").appendChild(card);
  (preset.holdings || [{}]).forEach((h) => addHolding(holdingsEl, h));
  updatePortfolioSubtotal(card);
  toggleAddPortfolioBtn();
}

function toggleAddPortfolioBtn() {
  document.getElementById("btn-add-portfolio").style.display =
    document.querySelectorAll(".portfolio-card").length >= 3 ? "none" : "";
}

// ── 입력 수집 ──────────────────────────────────────────────
function collectInput() {
  const portfolios = [];
  document.querySelectorAll(".portfolio-card").forEach((card) => {
    const holdings = [];
    card.querySelectorAll(".holding-row").forEach((r) => {
      const amount = Number(r.querySelector(".h-amount").value) || 0;
      if (amount <= 0) return;
      holdings.push({
        product_type: r.querySelector(".h-type").value,
        label: r.querySelector(".h-label").value,
        amount,
        annual_yield_pct: Number(r.querySelector(".h-yield").value) || 0,
      });
    });
    portfolios.push({ label: card.querySelector(".pf-name").value, holdings });
  });
  return {
    basic: {
      current_age: num("current_age"),
      target_age: num("target_age"),
      married: document.getElementById("married").checked,
      insurance_status: document.getElementById("insurance_status").value,
      public_pension_start_age: num("public_pension_start_age"),
      annual_living_expense: num("annual_living_expense"),
    },
    pension: {
      public_annual: num("public_annual"),
      private_annual: num("private_annual"),
      tax_free_annual: num("tax_free_annual"),
      housing_monthly: num("housing_monthly"),
    },
    property: {
      real_estate_tax_base: num("real_estate_tax_base"),
      rent_monthly: num("rent_monthly"),
    },
    portfolios,
  };
}

// ── API 호출 ──────────────────────────────────────────────
document.getElementById("btn-simulate").addEventListener("click", async () => {
  const errorBox = document.getElementById("error-box");
  errorBox.classList.add("hidden");
  const input = collectInput();
  if (input.portfolios.every((p) => p.holdings.length === 0)) {
    errorBox.textContent = "⚠️ 상품을 1개 이상 입력해 주세요 (평가금액과 수익률 필요).";
    errorBox.classList.remove("hidden");
    return;
  }
  const btn = document.getElementById("btn-simulate");
  btn.disabled = true; btn.textContent = "계산 중...";
  try {
    const res = await fetch("/api/simulate-portfolio", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(input),
    });
    if (!res.ok) {
      throw new Error(res.status === 422
        ? "입력값을 확인해 주세요 (나이 범위, 수익률 0~100%)."
        : "서버 오류 (" + res.status + ")");
    }
    lastResult = await res.json();
    renderResults(lastResult);
    document.getElementById("result-section").classList.remove("hidden");
    document.getElementById("result-section").scrollIntoView({ behavior: "smooth" });
  } catch (e) {
    errorBox.textContent = "⚠️ " + e.message;
    errorBox.classList.remove("hidden");
  } finally {
    btn.disabled = false; btn.textContent = "📊 비교 계산하기";
  }
});

// ── 결과 렌더링 ────────────────────────────────────────────
function renderResults(r) {
  const ps = r.portfolios;
  // 비교 표
  const head = `<tr><th>항목</th>${ps.map((p) => `<th>${escapeHtml(p.label)}</th>`).join("")}</tr>`;
  const rowsDef = [
    ["평가액 합계", (p) => formatWon(p.total_value)],
    ["연 금융소득", (p) => formatWon(p.financial_income_total)],
    ["금융소득종합과세", (p) => p.aggregate_subject
      ? `<span class="bad-text">⚠️ 대상</span>`
      : `<span class="good-text">✅ 비대상</span>`],
    [`${r.target_age}세 예상 세금(연)`, (p) => formatWon(p.target_snapshot.income_tax.total_tax)],
    [`${r.target_age}세 월 건보료`, (p) => formatWon(p.target_snapshot.health_insurance.monthly_premium)],
    ["피부양자 유지", (p) => p.target_snapshot.health_insurance.dependent_eligible
      ? "✅ 유지" : `<span class="bad-text">불가</span>`],
    [`${r.target_age}세 순현금흐름(연)`, (p) => formatWon(p.target_snapshot.net)],
    ["자산 소진", (p) => p.asset_depletion_age
      ? `<span class="bad-text">${p.asset_depletion_age}세</span>` : "90세까지 유지 ✅"],
    ["90세 자산 잔액", (p) => formatWon(p.cashflow_by_age[p.cashflow_by_age.length - 1].remaining_assets)],
  ];
  const body = rowsDef.map(([label, fn]) =>
    `<tr><td class="row-label">${label}</td>${ps.map((p) => `<td>${fn(p)}</td>`).join("")}</tr>`
  ).join("");
  document.getElementById("compare-table").innerHTML = head + body;

  // 금융소득 내역 카드
  document.getElementById("breakdowns").innerHTML = ps.map((p) => {
    const items = Object.entries(p.income_breakdown)
      .map(([k, v]) => `<li>${escapeHtml(k)}: <strong>${formatWon(v)}</strong></li>`).join("");
    const tf = p.tax_free_income > 0
      ? `<li class="good-text">분리과세·비과세(금융소득 제외): ${formatWon(p.tax_free_income)}</li>` : "";
    return `<div class="card"><h4>${escapeHtml(p.label)}</h4><ul class="breakdown">${items}${tf}</ul></div>`;
  }).join("");

  // 상세 드롭다운
  const pick = document.getElementById("detail-pick");
  pick.innerHTML = ps.map((p, i) => `<option value="${i}">${escapeHtml(p.label)}</option>`).join("");
  renderDetailTable(ps[0]);

  renderCharts(ps);

  document.getElementById("disclaimer-list").innerHTML =
    r.disclaimers.map((d) => `<li>${escapeHtml(d)}</li>`).join("");
}

const COLORS = ["#1e3a8a", "#059669", "#d97706"];

function renderCharts(ps) {
  const ages = ps[0].cashflow_by_age.map((r) => r.age);
  const makeDatasets = (field) => ps.map((p, i) => ({
    label: p.label,
    data: p.cashflow_by_age.map((r) => r[field]),
    borderColor: COLORS[i], backgroundColor: COLORS[i],
    tension: 0.2, pointRadius: 0, borderWidth: 2,
  }));
  const options = {
    responsive: true,
    interaction: { mode: "index", intersect: false },
    plugins: {
      tooltip: { callbacks: { label: (c) => `${c.dataset.label}: ${formatWon(c.parsed.y)}` } },
      legend: { labels: { boxWidth: 14 } },
    },
    scales: {
      x: { title: { display: true, text: "나이" } },
      y: { ticks: { callback: (v) => (v / 100_000_000).toFixed(1) + "억" } },
    },
  };
  if (chartNet) chartNet.destroy();
  chartNet = new Chart(document.getElementById("chart-net"),
    { type: "line", data: { labels: ages, datasets: makeDatasets("net") }, options });
  if (chartAssets) chartAssets.destroy();
  chartAssets = new Chart(document.getElementById("chart-assets"),
    { type: "line", data: { labels: ages, datasets: makeDatasets("remaining_assets") }, options });
}

function renderDetailTable(p) {
  const rows = p.cashflow_by_age.map((r) => `<tr>
    <td>${r.age}세</td><td>${r.status}</td><td>${formatWon(r.income)}</td>
    <td>${formatWon(r.tax)}</td><td>${formatWon(r.insurance)}</td>
    <td>${formatWon(r.net)}</td><td>${formatWon(r.remaining_assets)}</td></tr>`).join("");
  document.getElementById("detail-table").innerHTML =
    `<tr><th>나이</th><th>건보 자격</th><th>총수입</th><th>세금</th><th>건보료(연)</th><th>순현금흐름</th><th>자산 잔액</th></tr>` + rows;
}

document.getElementById("detail-pick").addEventListener("change", (e) => {
  if (lastResult) renderDetailTable(lastResult.portfolios[Number(e.target.value)]);
});

function escapeHtml(s) {
  return String(s).replace(/[&<>"]/g, (c) =>
    ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;" }[c]));
}

// ── 초기화 ────────────────────────────────────────────────
document.getElementById("btn-add-portfolio").addEventListener("click", () => addPortfolio());

(async function init() {
  attachMoneyDisplays();
  try {
    PRODUCT_TYPES = await (await fetch("/api/product-types")).json();
  } catch (e) {
    PRODUCT_TYPES = { domestic_stock: { label: "국내 주식 (배당주)", note: "" } };
  }
  // 예시 2개 프리필: 배당주 중심 vs ETF 중심
  addPortfolio({
    label: "배당주 중심",
    holdings: [
      { product_type: "domestic_stock", label: "고배당주", amount: 400000000, annual_yield_pct: 4.5 },
      { product_type: "reits", label: "리츠", amount: 200000000, annual_yield_pct: 6 },
    ],
  });
  addPortfolio({
    label: "ETF 중심",
    holdings: [
      { product_type: "domestic_equity_etf", label: "국내 주식형 ETF", amount: 400000000, annual_yield_pct: 1.5 },
      { product_type: "foreign_etf", label: "해외 상장 ETF", amount: 200000000, annual_yield_pct: 1.2 },
    ],
  });
})();

File: backend/portfolio.py
"""포트폴리오 설계 모드: 상품 구성 → 배당·이자 산출 → 종합과세/세금/건보료 역산.

사용자가 '무엇에 얼마를 어떤 수익률로' 담을지 입력하면, 상품 유형에 따라
세금 분류가 자동으로 결정된다. 기존 계산 엔진(tax.py, health_insurance.py)을
그대로 재사용한다. 설명서는 docs/07_포트폴리오_설계.md 참고.
"""

from typing import Literal

from pydantic import BaseModel, Field

from . import rules
from .health_insurance import calc_health_insurance
from .models import FinancialInput, IncomeInput, PensionInput, PropertyInput
from .tax import (calc_financial_income, calc_income_tax, classify_pension,
                  is_subject_to_aggregate_financial_tax)

# ── 상품 유형 등록표 ──────────────────────────────────────
# field: 산출 소득이 합산될 FinancialInput 필드 (None이면 분리과세·비과세 → 금융소득 제외)
# note: 화면 안내용 한 줄 설명
PRODUCT_TYPES: dict[str, dict] = {
    "domestic_stock": {
        "label": "국내 주식 (배당주)", "field": "dividend_domestic",
        "note": "배당금이 금융소득에 합산됩니다. 매매차익은 비과세입니다.",
    },
    "domestic_equity_etf": {
        "label": "국내 주식형 ETF", "field": "etf_domestic_equity_dist",
        "note": "분배금만 금융소득. 매매차익은 비과세라 종합과세·건보 위험이 가장 낮습니다.",
    },
    "domestic_other_etf": {
        "label": "기타 국내 ETF (해외지수·채권·리츠 등)",
        "field": "etf_domestic_other_dist_and_gain",
        "note": "분배금과 과세 대상 매매차익이 모두 금융소득에 합산됩니다.",
    },
    "foreign_stock": {
        "label": "해외 주식", "field": "dividend_foreign",
        "note": "배당금만 금융소득. 양도차익은 별도(양도세 22%)라 종합과세·건보에 거의 영향이 없습니다.",
    },
    "foreign_etf": {
        "label": "해외 상장 ETF", "field": "etf_foreign_dist",
        "note": "분배금만 금융소득. 양도차익은 별도(양도세 22%)입니다.",
    },
    "bond": {
        "label": "채권", "field": "interest",
        "note": "이자가 금융소득에 합산됩니다.",
    },
    "deposit": {
        "label": "예금·적금", "field": "interest",
        "note": "이자가 금융소득에 합산됩니다.",
    },
    "reits": {
        "label": "리츠 (REITs)", "field": "dividend_domestic",
        "note": "배당금이 금융소득에 합산됩니다.",
    },
    "separate_or_taxfree": {
        "label": "분리과세·비과세 (ISA·연금저축·IRP 등)", "field": None,
        "note": "이 수익은 금융소득·건보료 계산에서 제외되고 현금흐름에만 더해집니다.",
    },
}

PRODUCT_CODES = tuple(PRODUCT_TYPES.keys())


# ── 입력 모델 ─────────────────────────────────────────────

class PortfolioHolding(BaseModel):
    product_type: Literal[PRODUCT_CODES]  # type: ignore[valid-type]
    label: str = Field(default="", description="메모 (예: TIGER 미국S&P500)")
    amount: int = Field(ge=0, description="목표 나이 시점의 평가금액")
    annual_yield_pct: float = Field(
        ge=0, le=100, description="연 분배/배당/이자 수익률 (%)"
    )


class Portfolio(BaseModel):
    label: str = Field(description="포트폴리오 이름 (예: 배당주 중심)")
    holdings: list[PortfolioHolding] = Field(default_factory=list)


class PortfolioBasic(BaseModel):
    current_age: int = Field(ge=19, le=89)
    target_age: int = Field(ge=40, le=90, description="이 포트폴리오로 사는 시작 나이")
    married: bool = False
    insurance_status: Literal["직장", "지역", "피부양자"] = "지역"
    public_pension_start_age: int = Field(default=65, ge=50, le=80)
    annual_living_expense: int = Field(default=0, ge=0)


class PortfolioSimInput(BaseModel):
    basic: PortfolioBasic
    pension: PensionInput = PensionInput()
    property: PropertyInput = PropertyInput()
    portfolios: list[Portfolio] = Field(min_length=1, max_length=3)


# ── 핵심 계산 ─────────────────────────────────────────────

def portfolio_to_financial(holdings: list[PortfolioHolding]) -> dict:
    """보유 상품을 금융소득(FinancialInput) + 비과세 현금흐름 + 평가총액으로 환산."""
    fields = {
        "interest": 0, "dividend_domestic": 0, "dividend_foreign": 0,
        "etf_domestic_equity_dist": 0, "etf_domestic_other_dist_and_gain": 0,
        "etf_foreign_dist": 0, "capital_gain_foreign": 0,
    }
    tax_free_income = 0
    total_value = 0
    breakdown: dict[str, int] = {}

    for h in holdings:
        income = round(h.amount * h.annual_yield_pct / 100)
        total_value += h.amount
        meta = PRODUCT_TYPES[h.product_type]
        if meta["field"] is None:
            tax_free_income += income
        else:
            fields[meta["field"]] += income
        breakdown[meta["label"]] = breakdown.get(meta["label"], 0) + income

    return {
        "financial": FinancialInput(**fields),
        "tax_free_income": tax_free_income,
        "total_value": total_value,
        "breakdown": breakdown,
    }


def simulate_one_portfolio(basic: PortfolioBasic, pension: PensionInput,
                           prop: PropertyInput, portfolio: Portfolio) -> dict:
    mapped = portfolio_to_financial(portfolio.holdings)
    fin = mapped["financial"]
    fin_summary = calc_financial_income(fin)
    fin_total = fin_summary["financial_income_total"]
    judge = is_subject_to_aggregate_financial_tax(fin_total)

    prop_for_sim = PropertyInput(
        real_estate_tax_base=prop.real_estate_tax_base,
        rent_monthly=prop.rent_monthly,
        financial_assets=mapped["total_value"],
    )
    no_earned = IncomeInput()  # 포트폴리오 모드는 근로·사업소득 없음 (노후 자산 중심)
    rent_annual = prop.rent_monthly * 12

    rows = []
    assets = float(mapped["total_value"])
    living = basic.annual_living_expense
    depletion_age = None
    target_snapshot = None

    for age in range(basic.target_age, rules.SIM_END_AGE + 1):
        public_active = age >= basic.public_pension_start_age
        pen = classify_pension(pension, age, public_active, private_active=True)
        tax = calc_income_tax(no_earned, prop.rent_monthly, basic.married,
                              fin_summary, pen, earned_active=False)
        hi = calc_health_insurance(basic.insurance_status, fin_total,
                                   pen["public_annual"], no_earned,
                                   prop.rent_monthly, prop_for_sim,
                                   earned_active=False)
        income_total = (
            fin_total + fin_summary["foreign_capital_gain"]
            + mapped["tax_free_income"] + rent_annual
            + pen["public_annual"] + pen["private_annual"]
            + pen["tax_free_cashflow"]
        )
        net = income_total - tax["total_tax"] - hi["annual_premium"]
        assets = assets + (net - living)  # 잉여는 적립, 부족은 원금 인출 (성장 미가정)
        if living > 0 and assets < 0 and depletion_age is None:
            depletion_age = age
        if target_snapshot is None:
            target_snapshot = {"age": age, "income_tax": tax, "health_insurance": hi,
                               "income_total": round(income_total), "net": round(net)}
        rows.append({"age": age, "status": hi["status_applied"],
                     "income": round(income_total), "tax": tax["total_tax"],
                     "insurance": hi["annual_premium"], "net": round(net),
                     "remaining_assets": round(assets)})

    return {
        "label": portfolio.label,
        "total_value": mapped["total_value"],
        "financial_income_total": fin_total,
        "income_breakdown": mapped["breakdown"],
        "tax_free_income": mapped["tax_free_income"],
        "aggregate_subject": judge["subject"],
        "aggregate_margin": judge["margin"],
        "target_snapshot": target_snapshot,
        "cashflow_by_age": rows,
        "asset_depletion_age": depletion_age,
    }


DISCLAIMERS = [
    "본 결과는 단순화된 모델에 따른 참고용 추정치이며, 실제 세액·보험료와 다를 수 있습니다.",
    "수익률은 입력하신 가정치이며, 시장 상황에 따라 실제 배당·분배금은 달라집니다.",
    "해외 주식/ETF의 양도차익(매도 차익)은 이 모드에서 0으로 가정합니다(분배금·배당만 반영). "
    "양도차익은 종합과세·건보료에 거의 영향이 없으나, 양도세는 별도로 발생합니다.",
    "물가상승은 반영하지 않으며, 모든 금액은 현재 가치 기준입니다.",
]


def simulate_portfolios(data: PortfolioSimInput) -> dict:
    results = []
    for p in data.portfolios:
        results.append(
            simulate_one_portfolio(data.basic, data.pension, data.property, p)
        )
    return {"portfolios": results, "target_age": data.basic.target_age,
            "disclaimers": DISCLAIMERS}
